packages/library/src/providers/ModalProvider/ModalProvider.tsx

```diff
@@ -0,0 +1,132 @@
+import React, {
+    createContext,
+    Dispatch,
+    PropsWithChildren,
+    ReactNode,
+    RefObject,
+    SetStateAction,
+    useCallback,
+    useContext,
+    useEffect,
+    useMemo,
+    useRef,
+    useState,
+} from 'react';
+import { createPortal } from 'react-dom';
+import { useOnClickOutside } from 'usehooks-ts';
+import { useBreakpoint } from '@deriv/quill-design';
+
+type TModalState = {
+    // Add your modal states here
+    modalType?: string; // Just an example
+};
+
+type TModalContext = {
+    getModalState: <T extends keyof TModalState>(key: T) => TModalState[T];
+    hide: () => void;
+    isOpen: boolean;
+    modalState?: Map<keyof TModalState, TModalState[keyof TModalState]>;
+    setModalOptions: Dispatch<SetStateAction<TModalOptions>>;
+    setModalState: <T extends keyof TModalState>(key: T, value: TModalState[T]) => void;
+    show: (ModalContent: ReactNode, options?: TModalOptions) => void;
+};
+
+type TModalOptions = {
+    defaultRootId?: 'wallets_modal_root' | 'wallets_modal_show_header_root';
```

Comment:
```suggestion
    defaultRootId?: 'wallets_modal_root' | 'wallets_modal_show_header_root';
```
Nit: maybe we should rename this? 'v2_modal_root' ? Gotta update the `index.html `as well if you decide to do this

Reply:
If i add one more there will be 3 modal roots doing almost the same thing. I think we can keep this for now and when wallets are using this we can rename this and the one in index.html

Reply:
the idea is to target the same modal root for all the v2 providers. so technically the one in wallets need to be renamed as well - because the idea is to use one modal root for all apps (basically we are making wallets initiative universal now)